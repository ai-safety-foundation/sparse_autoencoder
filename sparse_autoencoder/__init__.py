"""Sparse Autoencoder Library."""
from sparse_autoencoder.activation_resampler.activation_resampler import ActivationResampler
from sparse_autoencoder.activation_store.disk_store import DiskActivationStore
from sparse_autoencoder.activation_store.list_store import ListActivationStore
from sparse_autoencoder.activation_store.tensor_store import TensorActivationStore
from sparse_autoencoder.autoencoder.model import SparseAutoencoder
from sparse_autoencoder.loss.abstract_loss import LossLogType, LossReductionType
from sparse_autoencoder.loss.decoded_activations_l2 import L2ReconstructionLoss
from sparse_autoencoder.loss.learned_activations_l1 import LearnedActivationsL1Loss
from sparse_autoencoder.loss.reducer import LossReducer
from sparse_autoencoder.metrics.resample.neuron_activity_metric import NeuronActivityMetric
from sparse_autoencoder.metrics.train.capacity import CapacityMetric
from sparse_autoencoder.metrics.train.feature_density import TrainBatchFeatureDensityMetric
from sparse_autoencoder.optimizer.adam_with_reset import AdamWithReset
from sparse_autoencoder.source_data.pretokenized_dataset import PreTokenizedDataset
from sparse_autoencoder.source_data.text_dataset import TextDataset
from sparse_autoencoder.train.pipeline import Pipeline


__all__ = [
    "ActivationResampler",
    "AdamWithReset",
    "CapacityMetric",
    "DiskActivationStore",
    "L2ReconstructionLoss",
    "LearnedActivationsL1Loss",
    "ListActivationStore",
    "LossLogType",
    "LossReducer",
    "LossReductionType",
    "NeuronActivityMetric",
    "Pipeline",
    "PreTokenizedDataset",
    "SparseAutoencoder",
    "TensorActivationStore",
    "TextDataset",
    "TrainBatchFeatureDensityMetric",
]
