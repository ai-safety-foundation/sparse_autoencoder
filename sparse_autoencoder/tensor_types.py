"""Tensor Types.

Tensor types with axis labels. Note that this uses the `jaxtyping` library, which works with PyTorch
tensors as well. Note also that shape sizes are included in the docstrings as well as in the types
as this is needed for IDEs such as VSCode to provide code hints.
"""
from enum import auto
from typing import TypeAlias

from jaxtyping import Float, Int
from strenum import LowercaseStrEnum
from torch import Tensor


class Axis(LowercaseStrEnum):
    """Tensor axis names.

    Used to annotate tensor types.

    Example:
        When used directly it prints a string:

        >>> print(Axis.INPUT_OUTPUT_FEATURE)
        input_output_feature

        The primary use is to annotate tensor types:

        >>> from jaxtyping import Float
        >>> from torch import Tensor
        >>> from typing import TypeAlias
        >>> batch: TypeAlias = Float[Tensor, Axis.dims(Axis.BATCH, Axis.INPUT_OUTPUT_FEATURE)]
        >>> print(batch)
        <class 'jaxtyping.Float[Tensor, 'batch input_output_feature']'>

        You can also join multiple axis together to represent the dimensions of a tensor:

        >>> print(Axis.dims(Axis.BATCH, Axis.INPUT_OUTPUT_FEATURE))
        batch input_output_feature
    """

    # Batches
    SOURCE_DATA_BATCH = auto()
    """Batch of prompts used to generate source model activations."""

    BATCH = auto()
    """Batch of items that the SAE is being trained on."""

    ITEMS = auto()
    """Arbitrary number of items."""

    # Features
    INPUT_OUTPUT_FEATURE = auto()
    """Input or output feature (e.g. feature in activation vector from source model)."""

    LEARNT_FEATURE = auto()
    """Learn feature (e.g. feature in learnt activation vector)."""

    DEAD_FEATURE = auto()
    """Dead feature."""

    ALIVE_FEATURE = auto()
    """Alive feature."""

    # Feature indices
    LEARNT_FEATURE_IDX = auto()

    # Other
    POSITION = auto()
    """Token position."""

    SINGLE_ITEM = ""
    """Single item axis."""

    ANY = "*any"
    """Any number of axis."""

    @staticmethod
    def dims(*axis: "Axis") -> str:
        """Join multiple axis together, to represent the dimensions of a tensor.

        Example:
            >>> print(Axis.dims(Axis.BATCH, Axis.INPUT_OUTPUT_FEATURE))
            batch input_output_feature

        Args:
            *axis: Axis to join.

        Returns:
            Joined axis string.
        """
        return " ".join(axis)


# Activation vectors
InputOutputActivationVector: TypeAlias = Float[Tensor, Axis.INPUT_OUTPUT_FEATURE]
"""Input/output activation vector.

This is either a input activation vector from the source model, or a decoded activation vector
from the autoencoder.
"""

LearntActivationVector: TypeAlias = Float[Tensor, Axis.LEARNT_FEATURE]
"""Learned activation vector.

Activation vector from the hidden (learnt) layer of the autoencoder. Typically this is larger than
the input/output activation vector.
"""

# Activation batches/stores
StoreActivations: TypeAlias = Float[Tensor, Axis.dims(Axis.ITEMS, Axis.INPUT_OUTPUT_FEATURE)]
"""Store of activation vectors.

This is used to store large numbers of activation vectors from the source model.
"""

SourceModelActivations: TypeAlias = Float[Tensor, Axis.dims(Axis.ANY, Axis.INPUT_OUTPUT_FEATURE)]
"""Source model activations.

Can have any number of proceeding dimensions (e.g. an attention head may generate activations of
shape (batch_size, num_heads, seq_len, feature_dim).
"""

InputOutputActivationBatch: TypeAlias = Float[
    Tensor, Axis.dims(Axis.BATCH, Axis.INPUT_OUTPUT_FEATURE)
]
"""Input/output activation batch.

This is either a batch of input activation vectors from the source model, or a batch of decoded
activation vectors from the autoencoder.

Shape (batch, input_output_feature)
"""

LearnedActivationBatch: TypeAlias = Float[Tensor, Axis.dims(Axis.BATCH, Axis.LEARNT_FEATURE)]
"""Learned activation batch.

This is a batch of activation vectors from the hidden (learnt) layer of the autoencoder. Typically
the feature dimension is larger than the input/output activation vector.
"""

# Statistics
TrainBatchStatistic: TypeAlias = Float[Tensor, Axis.BATCH]
"""Train batch statistic.

Contains one scalar value per item in the batch.
"""

ValidationStatistics: TypeAlias = Float[Tensor, Axis.ITEMS]
"""Validation statistics.

Contains one scalar value per item in the validation set.
"""

# Weights and biases
EncoderWeights: TypeAlias = Float[Tensor, Axis.dims(Axis.LEARNT_FEATURE, Axis.INPUT_OUTPUT_FEATURE)]
"""Encoder weights.

These weights are part of the encoder module of the autoencoder, responsible for decompressing the
input data (activations from a source model) into a higher-dimensional representation.

The dictionary vectors (basis vectors in the learnt feature space), they can be thought of as
columns of this weight matrix, where each column corresponds to a particular feature in the
lower-dimensional space. The sparsity constraint (hopefully) enforces that they respond relatively
strongly to only a small portion of possible input vectors.

Shape: (learnt_feature_dim, input_output_feature_dim)
"""

DecoderWeights: TypeAlias = Float[Tensor, Axis.dims(Axis.INPUT_OUTPUT_FEATURE, Axis.LEARNT_FEATURE)]
"""Decoder weights.

These weights form the decoder part of the autoencoder, which aims to reconstruct the original input
data from the decompressed representation created by the encoder.

Viewing the dictionary vectors in the context of reconstruction, they can be thought of as rows in
this weight matrix.

Shape: (input_output_feature_dim, learnt_feature_dim)
"""

# Weights and biases updated
NeuronActivity: TypeAlias = Int[Tensor, Axis.LEARNT_FEATURE]
"""Neuron activity.

Number of times each neuron has fired (since the last reset).
"""

InputOutputNeuronIndices: TypeAlias = Int[Tensor, Axis.INPUT_OUTPUT_FEATURE]
"""Input/output neuron indices."""

LearntNeuronIndices: TypeAlias = Int[Tensor, Axis.LEARNT_FEATURE_IDX]
"""Learnt neuron indices."""

SampledDeadNeuronInputs: TypeAlias = Float[
    Tensor, Axis.dims(Axis.DEAD_FEATURE, Axis.INPUT_OUTPUT_FEATURE)
]
"""Sampled dead neuron inputs.

Shape: (dead_feature, input_output_feature)
"""

AliveEncoderWeights: TypeAlias = Float[Tensor, Axis.dims(Axis.LEARNT_FEATURE, Axis.ALIVE_FEATURE)]
"""Alive encoder weights."""

DeadEncoderNeuronWeightUpdates: TypeAlias = Float[
    Tensor, Axis.dims(Axis.DEAD_FEATURE, Axis.INPUT_OUTPUT_FEATURE)
]
"""Dead encoder neuron weight updates.

Shape (learnt_feature, dead_feature)
"""

DeadEncoderNeuronBiasUpdates: TypeAlias = Float[Tensor, Axis.DEAD_FEATURE]
"""Dead encoder neuron bias updates."""

DeadDecoderNeuronWeightUpdates: TypeAlias = Float[
    Tensor, Axis.dims(Axis.INPUT_OUTPUT_FEATURE, Axis.DEAD_FEATURE)
]
"""Dead decoder neuron weight updates.

Shape (dead_feature, learnt_feature)
"""

# Other
BatchTokenizedPrompts: TypeAlias = Int[Tensor, Axis.dims(Axis.SOURCE_DATA_BATCH, Axis.POSITION)]
"""Batch of tokenized prompts."""

ItemTensor: TypeAlias = Float[Tensor, Axis.SINGLE_ITEM]
"""Single element item tensor."""
