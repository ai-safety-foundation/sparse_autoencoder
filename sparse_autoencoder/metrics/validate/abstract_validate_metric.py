"""Abstract metric classes."""
from abc import ABC, abstractmethod
from dataclasses import dataclass
from typing import Any

from sparse_autoencoder.tensor_types import ValidationStatistics


@dataclass
class ValidationMetricData:
    """Validation metric data."""

    source_model_loss: ValidationStatistics

    source_model_loss_with_reconstruction: ValidationStatistics

    source_model_loss_with_zero_ablation: ValidationStatistics


class AbstractValidationMetric(ABC):
    """Abstract validation metric."""

    @abstractmethod
    def calculate(self, data: ValidationMetricData) -> dict[str, Any]:
        """Calculate any metrics."""
