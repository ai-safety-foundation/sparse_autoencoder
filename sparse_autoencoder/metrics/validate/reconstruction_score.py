"""Reconstruction score metric."""
from jaxtyping import Float, Int64
from pydantic import PositiveInt, validate_call
import torch
from torch import Tensor
from torchmetrics import Metric

from sparse_autoencoder.tensor_types import Axis


class ReconstructionScoreMetric(Metric):
    r"""Model reconstruction score.

    Creates a score that measures how well the model can reconstruct the data.

    $$
    \begin{align*}
        v &= \text{number of validation items} \\
        l \in{\mathbb{R}^v} &= \text{loss with no changes to the source model} \\
        l_\text{recon} \in{\mathbb{R}^v} &= \text{loss with reconstruction} \\
        l_\text{zero} \in{\mathbb{R}^v} &= \text{loss with zero ablation} \\
        s &= \text{reconstruction score} \\
        s_\text{itemwise} &= \frac{l_\text{zero} - l_\text{recon}}{l_\text{zero} - l} \\
        s &= \sum_{i=1}^v s_\text{itemwise} / v
    \end{align*}
    $$

    Example:
        >>> metric = ReconstructionScoreMetric(num_components=1)
        >>> source_model_loss=torch.tensor([2.0, 2.0, 2.0])
        >>> source_model_loss_with_reconstruction=torch.tensor([3.0, 3.0, 3.0])
        >>> source_model_loss_with_zero_ablation=torch.tensor([5.0, 5.0, 5.0])
        >>> metric.forward(
        ...     source_model_loss=source_model_loss,
        ...     source_model_loss_with_reconstruction=source_model_loss_with_reconstruction,
        ...     source_model_loss_with_zero_ablation=source_model_loss_with_zero_ablation
        ... )
        tensor(0.6667)
    """

    # Torchmetrics settings
    is_differentiable: bool | None = False
    full_state_update: bool | None = True

    # State
    source_model_loss: Float[Tensor, Axis.COMPONENT_OPTIONAL]
    source_model_loss_with_zero_ablation: Float[Tensor, Axis.COMPONENT_OPTIONAL]
    source_model_loss_with_reconstruction: Float[Tensor, Axis.COMPONENT_OPTIONAL]
    num_activation_vectors: Int64[Tensor, Axis.SINGLE_ITEM]

    @validate_call
    def __init__(self, num_components: PositiveInt = 1) -> None:
        """Initialise the metric."""
        super().__init__()

        self.add_state(
            "source_model_loss", default=torch.zeros(num_components), dist_reduce_fx="sum"
        )
        self.add_state(
            "source_model_loss_with_zero_ablation",
            default=torch.zeros(num_components),
            dist_reduce_fx="sum",
        )
        self.add_state(
            "source_model_loss_with_reconstruction",
            default=torch.zeros(num_components),
            dist_reduce_fx="sum",
        )

    def update(
        self,
        source_model_loss: Float[Tensor, Axis.COMPONENT_OPTIONAL],
        source_model_loss_with_reconstruction: Float[Tensor, Axis.COMPONENT_OPTIONAL],
        source_model_loss_with_zero_ablation: Float[Tensor, Axis.COMPONENT_OPTIONAL],
        component_idx: int = 0,
    ) -> None:
        """Update the metric state.

        Args:
            source_model_loss: Loss with no changes to the source model.
            source_model_loss_with_reconstruction: Loss with SAE reconstruction.
            source_model_loss_with_zero_ablation: Loss with zero ablation.
            component_idx: Component idx.
        """
        self.source_model_loss[component_idx] += source_model_loss.sum()
        self.source_model_loss_with_zero_ablation[
            component_idx
        ] += source_model_loss_with_zero_ablation.sum()
        self.source_model_loss_with_reconstruction[
            component_idx
        ] += source_model_loss_with_reconstruction.sum()

    def compute(
        self,
    ) -> Float[Tensor, Axis.COMPONENT_OPTIONAL]:
        """Compute the metric."""
        zero_ablate_loss_minus_reconstruction_loss: Float[Tensor, Axis.COMPONENT_OPTIONAL] = (
            self.source_model_loss_with_zero_ablation - self.source_model_loss_with_reconstruction
        )

        zero_ablate_loss_minus_default_loss: Float[Tensor, Axis.COMPONENT_OPTIONAL] = (
            self.source_model_loss_with_zero_ablation - self.source_model_loss
        )

        return zero_ablate_loss_minus_reconstruction_loss / zero_ablate_loss_minus_default_loss
