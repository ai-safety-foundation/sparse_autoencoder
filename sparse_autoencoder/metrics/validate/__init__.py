"""Validate step metrics."""
from sparse_autoencoder.metrics.validate.abstract_validate_metric import AbstractValidationMetric


__all__ = ["AbstractValidationMetric"]
