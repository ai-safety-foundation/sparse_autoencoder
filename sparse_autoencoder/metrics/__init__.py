"""Metrics."""
from sparse_autoencoder.metrics.generate import AbstractGenerateMetric
from sparse_autoencoder.metrics.train import AbstractTrainMetric
from sparse_autoencoder.metrics.validate import AbstractValidationMetric


__all__ = [
    "AbstractGenerateMetric",
    "AbstractTrainMetric",
    "AbstractValidationMetric",
]
