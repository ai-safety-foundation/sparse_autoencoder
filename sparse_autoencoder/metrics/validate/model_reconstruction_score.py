"""Model reconstruction score."""
from typing import TYPE_CHECKING, Any

from sparse_autoencoder.metrics.validate.abstract_validate_metric import (
    AbstractValidationMetric,
    ValidationMetricData,
)


if TYPE_CHECKING:
    from sparse_autoencoder.tensor_types import ValidationStatistics


class ModelReconstructionScore(AbstractValidationMetric):
    r"""Model reconstruction score.

    Creates a score that measures how well the model can reconstruct the data.

    $$
    \begin{align*}
        v &= \text{number of validation items} \\
        l \in{\mathbb{R}^v} &= \text{loss with no changes to the source model} \\
        l_\text{recon} \in{\mathbb{R}^v} &= \text{loss with reconstruction} \\
        l_\text{zero} \in{\mathbb{R}^v} &= \text{loss with zero ablation} \\
        s &= \text{reconstruction score} \\
        s_\text{itemwise} &= \frac{l_\text{zero} - l_\text{recon}}{l_\text{zero} - l} \\
        s &= \sum_{i=1}^v s_\text{itemwise} / v
    \end{align*}
    $$
    """

    def calculate(self, data: ValidationMetricData) -> dict[str, Any]:
        """Calculate the model reconstruction score.

        Example:
            >>> from sparse_autoencoder.tensor_types import ValidationStatistics
            >>> data = ValidationMetricData(
            ...     source_model_loss=ValidationStatistics([2.0, 2.0, 2.0]),
            ...     source_model_loss_with_reconstruction=ValidationStatistics([3.0, 3.0, 3.0]),
            ...     source_model_loss_with_zero_ablation=ValidationStatistics([5.0, 5.0, 5.0])
            ... )
            >>> metric = ModelReconstructionScore()
            >>> result = metric.calculate(data)
            >>> round(result['model_reconstruction_score'], 3)
            0.667

        Args:
            data: Validation data.

        Returns:
            Model reconstruction score.
        """
        # Return no statistics if the data is empty (e.g. if we're at the very end of training)
        if data.source_model_loss.numel() == 0:
            return {}

        # Calculate the reconstruction score
        zero_ablate_loss_minus_default_loss: ValidationStatistics = (
            data.source_model_loss_with_zero_ablation - data.source_model_loss
        )
        zero_ablate_loss_minus_reconstruction_loss: ValidationStatistics = (
            data.source_model_loss_with_zero_ablation - data.source_model_loss_with_reconstruction
        )
        model_reconstruction_score_itemwise: ValidationStatistics = (
            zero_ablate_loss_minus_reconstruction_loss / zero_ablate_loss_minus_default_loss
        )
        model_reconstruction_score: float = model_reconstruction_score_itemwise.mean().item()

        # Get the other metrics
        validation_baseline_loss: float = data.source_model_loss.mean().item()
        validation_loss_with_reconstruction: float = (
            data.source_model_loss_with_reconstruction.mean().item()
        )
        validation_loss_with_zero_ablation: float = (
            data.source_model_loss_with_zero_ablation.mean().item()
        )

        return {
            "validation_baseline_loss": validation_baseline_loss,
            "validation_loss_with_reconstruction": validation_loss_with_reconstruction,
            "validation_loss_with_zero_ablation": validation_loss_with_zero_ablation,
            "model_reconstruction_score": model_reconstruction_score,
        }
