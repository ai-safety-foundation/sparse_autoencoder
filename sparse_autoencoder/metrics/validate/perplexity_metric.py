"""Doing a periodic evaluation of the autoencoder."""
from collections.abc import Callable
from typing import Any

import torch
from torch import Tensor
from transformer_lens.hook_points import HookPoint

from sparse_autoencoder.autoencoder.abstract_autoencoder import \
    AbstractAutoencoder
from sparse_autoencoder.metrics.validate.abstract_validate_metric import (
    AbstractValidationMetric, ValidationMetricContext)
from sparse_autoencoder.train.utils import get_model_device

N_BATCHES = 100


def make_autoencoder_hook(
    autoencoder: AbstractAutoencoder
) -> Callable[[Tensor, HookPoint], Tensor]:
    """Make the hook that will go into the forward pass of the model."""

    def autoencoder_hook(input_activations: Tensor, hook: HookPoint) -> Tensor:  # noqa: ARG001
        """Hook to get the activations from the autoencoder."""
        to_sae_activations = input_activations.reshape(-1, autoencoder.n_input_features)
        _, reconstructed_activations = autoencoder(to_sae_activations)
        return reconstructed_activations.reshape(input_activations.shape)

    return autoencoder_hook


class PerplexityMetric(AbstractValidationMetric):
    """Calculate the change in perplexity when using the autoencoder in the original model."""

    def calculate(self, context: ValidationMetricContext) -> dict[str, Any]:
        """Run a perplexity evaluation of the model with and without the autoencoder."""
        total_standard_loss = 0.0
        total_sae_loss = 0.0
        total_kl_divergence = 0.0
        batches = 0

        device = get_model_device(context.source_model)

        with torch.no_grad():
            for batch_raw in context.dataset:
                batch = torch.tensor(batch_raw["input_ids"]).to(device=device).to(torch.int64)[:2]
                print(batch.shape)

                # Get the activations from the model
                sae_hook = make_autoencoder_hook(context.autoencoder)
                # standard_loss2 = context.source_model.run_with_hooks(batch, return_type="loss")
                # sae_loss2 = context.source_model.run_with_hooks(
                #     batch, return_type="loss", fwd_hooks=[(context.hook_point, sae_hook)]
                # )

                standard_logits = context.source_model.run_with_hooks(batch, return_type="logits")
                sae_logits = context.source_model.run_with_hooks(
                    batch, return_type="logits", fwd_hooks=[(context.hook_point, sae_hook)]
                )

                kl_divergence = torch.nn.functional.kl_div(
                    torch.nn.functional.log_softmax(standard_logits, dim=-1),
                    torch.nn.functional.softmax(sae_logits, dim=-1),
                    reduction="batchmean",
                )
                kl_divergence = kl_divergence / batch.shape[0]
                # loss should be equal to the negative log likelihood
                sae_logit_outputs = sae_logits[:, :-1].reshape(-1, sae_logits.shape[-1])
                standard_logit_outputs = standard_logits[:, :-1].reshape(-1, standard_logits.shape[-1])
                # sae_log_probs = torch.nn.functional.log_softmax(sae_logit_outputs, dim=-1)
                standard_log_probs = torch.nn.functional.log_softmax(standard_logit_outputs, dim=-1)

                targets = batch[1:]
                criterion = torch.nn.CrossEntropyLoss()
                nll_loss = torch.nn.NLLLoss()
                targets2 = torch.tensor(targets[0].item()).to(device=device).to(torch.int64).unsqueeze(0)
                print(criterion(sae_logit_outputs, targets2))
                print(criterion(sae_logit_outputs, targets))

                sae_loss = criterion(sae_logit_outputs, targets)
                standard_loss = torch.nn.functional.nll_loss(
                    standard_log_probs, targets, reduction="mean"
                )
                breakpoint()

                total_standard_loss += standard_loss.item()
                total_sae_loss += sae_loss.item()

                total_kl_divergence += kl_divergence.item()

                batches += 1
                if batches >= N_BATCHES:
                    break

        return {
            "Standard loss": total_standard_loss / batches,
            "SAE loss": total_sae_loss / batches,
            "KL divergence": total_kl_divergence / batches,
        }
