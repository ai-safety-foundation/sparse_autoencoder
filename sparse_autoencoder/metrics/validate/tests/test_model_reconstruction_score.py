"""Test the model reconstruction score metric."""

import pytest

from sparse_autoencoder.metrics.validate.abstract_validate_metric import ValidationMetricData
from sparse_autoencoder.metrics.validate.model_reconstruction_score import ModelReconstructionScore
from sparse_autoencoder.tensor_types import ValidationStatistics


def test_model_reconstruction_score_empty_data() -> None:
    """Test the model reconstruction score with empty data.

    This test validates that the method returns an empty dictionary when no data
    is provided (i.e., at the end of training or in similar scenarios).
    """
    data = ValidationMetricData(
        source_model_loss=ValidationStatistics([]),
        source_model_loss_with_reconstruction=ValidationStatistics([]),
        source_model_loss_with_zero_ablation=ValidationStatistics([]),
    )
    metric = ModelReconstructionScore()
    result = metric.calculate(data)
    assert result == {}


@pytest.mark.parametrize(
    ("data", "expected_score"),
    [
        (
            ValidationMetricData(
                source_model_loss=ValidationStatistics([3.0, 3.0, 3.0]),
                source_model_loss_with_reconstruction=ValidationStatistics([3.0, 3.0, 3.0]),
                source_model_loss_with_zero_ablation=ValidationStatistics([4.0, 4.0, 4.0]),
            ),
            1.0,
        ),
        (
            ValidationMetricData(
                source_model_loss=ValidationStatistics([0.5, 1.5, 2.5]),
                source_model_loss_with_reconstruction=ValidationStatistics([1.5, 2.5, 3.5]),
                source_model_loss_with_zero_ablation=ValidationStatistics([8.0, 7.0, 4.0]),
            ),
            0.67,
        ),
    ],
)
def test_model_reconstruction_score_various_data(
    data: ValidationMetricData, expected_score: float
) -> None:
    """Test the model reconstruction score with various data inputs.

    This test uses parameterization to check the model reconstruction score
    calculation for different sets of input data.
    """
    metric = ModelReconstructionScore()
    result = metric.calculate(data)
    assert round(result["model_reconstruction_score"], 2) == expected_score
