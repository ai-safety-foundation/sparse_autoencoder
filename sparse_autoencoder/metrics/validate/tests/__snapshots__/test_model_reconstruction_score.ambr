# serializer version: 1
# name: test_weights_biases_log_matches_snapshot
  list([
    dict({
      'component_0/validate/reconstruction_score/baseline_loss': tensor(0.3800),
      'component_1/validate/reconstruction_score/baseline_loss': tensor(0.5251),
      'component_2/validate/reconstruction_score/baseline_loss': tensor(0.4923),
      'component_3/validate/reconstruction_score/baseline_loss': tensor(0.4598),
      'component_4/validate/reconstruction_score/baseline_loss': tensor(0.4281),
      'component_5/validate/reconstruction_score/baseline_loss': tensor(0.4961),
      'validate/reconstruction_score/baseline_loss/component_mean': tensor(0.4636),
    }),
    dict({
      'component_0/validate/reconstruction_score/loss_with_reconstruction': tensor(0.6111),
      'component_1/validate/reconstruction_score/loss_with_reconstruction': tensor(0.5219),
      'component_2/validate/reconstruction_score/loss_with_reconstruction': tensor(0.4063),
      'component_3/validate/reconstruction_score/loss_with_reconstruction': tensor(0.6497),
      'component_4/validate/reconstruction_score/loss_with_reconstruction': tensor(0.4929),
      'component_5/validate/reconstruction_score/loss_with_reconstruction': tensor(0.3723),
      'validate/reconstruction_score/loss_with_reconstruction/component_mean': tensor(0.5090),
    }),
    dict({
      'component_0/validate/reconstruction_score/loss_with_zero_ablation': tensor(0.2891),
      'component_1/validate/reconstruction_score/loss_with_zero_ablation': tensor(0.3879),
      'component_2/validate/reconstruction_score/loss_with_zero_ablation': tensor(0.5850),
      'component_3/validate/reconstruction_score/loss_with_zero_ablation': tensor(0.4740),
      'component_4/validate/reconstruction_score/loss_with_zero_ablation': tensor(0.5452),
      'component_5/validate/reconstruction_score/loss_with_zero_ablation': tensor(0.3733),
      'validate/reconstruction_score/loss_with_zero_ablation/component_mean': tensor(0.4424),
    }),
    dict({
      'component_0/validate/reconstruction_score': tensor(3.5422),
      'component_1/validate/reconstruction_score': tensor(0.9767),
      'component_2/validate/reconstruction_score': tensor(1.9278),
      'component_3/validate/reconstruction_score': tensor(-12.3338),
      'component_4/validate/reconstruction_score': tensor(0.4468),
      'component_5/validate/reconstruction_score': tensor(-0.0081),
      'validate/reconstruction_score/component_mean': tensor(-0.9081),
    }),
  ])
# ---
